lib/propolis/src/hw/qemu/fwcfg.rs

```diff
@@ -1057,6 +1057,158 @@ mod test {
 pub mod formats {
     use super::Entry;
     use crate::hw::pci;
+    use zerocopy::{Immutable, IntoBytes};
+
+    /// A type for a range described in an E820 map entry.
+    ///
+    /// This is canonically defined as the ACPI "Address Range Types", though we
+    /// only define the types we use, which are a subset of the types that EDK2
+    /// is known to care about, which itself is a subset of types that ACPI and
+    /// OEMs define or guest OSes may care about.
+    #[derive(IntoBytes, Immutable)]
+    #[repr(u32)]
+    enum EfiAcpiMemoryType {
+        Memory = 1,
+        Reserved = 2,
+        // For reference, though these types are unused.
+        // Acpi = 3,
+        // Nvs = 4,
+    }
+
+    /// One address/length/type entry in the E820 map.
+    ///
+    /// This is... almost defined by ACPI's "Address Range Descriptor Structure"
+    /// table, under "INT 15H, E820". Critically, ACPI defines this structure
+    /// with an additional "Extended Attributes" field which EDK2 does not know
+    /// about and so we do not provide. Consequently the size of this struct is
+    /// 20 bytes as defined in `OvmfPkg/Include/IndustryStandard/E820.h` rather
+    /// than the ACPI definition's 24 bytes.
+    #[derive(IntoBytes, Immutable)]
+    #[repr(C, packed)]
+    struct E820Entry64 {
+        base_addr: u64,
+        length: u64,
+        ty: EfiAcpiMemoryType,
+    }
+
+    /// A list of E820 memory map entries.
+    ///
+    /// This is not defined by ACPI, but is an EDK2 implementation of a QMEU
+    /// construct to communicate an E820 map to the firmware. It is parsed by
+    /// EDK2 and added to its EFI memory map; it is not, itself, the memory map
+    /// that OVMF presents via UEFI services. It is not required to be sorted,
+    /// and EDK2 ignores entries starting below 4 GiB. Adding additional
+    /// low-memory entries is not harmful, but not valuable to EDK2 either.
+    pub struct E820Table(Vec<E820Entry64>);
```

Comment:
I'd be tempted to just publicly expose the `E820Entry64` and `EfiAcpiMemoryType` types, and providing a `From <&[E820Entry64]> for Entry` impl for the conversion, since `E820Table` isn't really providing much in the way of value-add over just shimming the Vec methods.  Up to you. *shrug*

Reply:
i should probably examine my dislike of `From`/`Into` one day, but i wonder if some of it is finding this easier to search for..